File: part1/task34_.py
# половина тестов не проходит по времени
import sys
input = sys.stdin.readline

def check(d):
    l = -1
    r = n
    while l+1 < r:
        mid = (l+r) // 2
        if a[mid] == d:
            return d
        if a[mid] < d:
            l = mid
        else:
            r = mid
    if r == n:
        return "NO SOLUTION"
    else:
        return a[l+1]

n, m = map(int, input().split())
a = list(map(int, input().split()))

for _ in range(m):
    print(check(int(input())))

File: part1/task34.cpp
#include <iostream>
using namespace std;

int n;

void check(int d, int a[]) {
    int l = -1, r = n, mid;
    while ((l+1) < r) {
        mid = (l+r) / 2;
        if (a[mid] == d) {
			cout << d << '\n'; return ;
		}
        if (a[mid] < d) {
			l = mid;
		}
        else {
			r = mid;
		}
	}
    if (r == n)
        cout << "NO SOLUTION\n";
    else
        cout << a[l+1] << '\n';
	return ;
}

int main() {
	int m, t; cin >> n; cin >> m;
	int a[n];
	for (int i = 0; i < n; i++) {
		cin >> t; a[i] = t;
	}
	for (int i = 0; i < m; i++) {
		cin >> t; check(t, a);
	}
}